components/service/experiments/src/main/java/mozilla/components/service/experiments/util/VersionString.kt

```diff
@@ -0,0 +1,73 @@
+package mozilla.components.service.experiments.util
+
+import java.lang.IllegalArgumentException
+import kotlin.math.max
+
+internal class VersionString(
+    val version: String
+) : Comparable<VersionString> {
+    companion object {
+        // Validates that the version matches expected version formatting
+        // This matches any number of digits separated by dots.
+        // Valid examples:
+        //  1.0.1
+        //  1.0
+        //  10.4.123545
+        // Invalid examples:
+        // a.b.c
+        // Not.a.version
+        // 12.1.1A
+        val VERSION_REGEX = "[0-9]+(\\.[0-9]+)*".toRegex()
+    }
+
+    @Suppress("ComplexMethod")
+    override fun compareTo(other: VersionString): Int {
+        if (!version.matches(VERSION_REGEX)) {
+            throw IllegalArgumentException("Unexpected format in VersionString")
+        }
+        if (!other.version.matches(VERSION_REGEX)) {
+            throw IllegalArgumentException("Unexpected format in VersionString")
+        }
+        val thisParts = version.split(".")
+        val otherParts = other.version.split(".")
+        val length = max(thisParts.count(), otherParts.count())
+        for (i in 0..length) {
+            val thisPart = if (i < thisParts.count()) {
```

Comment:
Non-blocking feedback: filling the `*Parts` lists up with `0`s before the loop would probably make for more readable code.

Reply:
Added a comment for now, to better explain what's going on.